src/OrleansServiceBus/Providers/Streams/EventHub/EventHubQueueCache.cs

```diff
@@ -2,35 +2,96 @@
         where TCachedMessage : struct
     {
+        protected readonly int defaultMaxAddCount;
+        protected readonly PooledQueueCache<EventData, TCachedMessage> cache;
+        private readonly CachePressureMonitor cachePressureMonitor;
+
         protected IStreamQueueCheckpointer<string> Checkpointer { private set; get; }
 
-        protected EventHubQueueCache(IStreamQueueCheckpointer<string> checkpointer, ICacheDataAdapter<EventData, TCachedMessage> cacheDataAdapter, ICacheDataComparer<TCachedMessage> comparer)
-            : base(cacheDataAdapter, comparer)
+        protected EventHubQueueCache(int defaultMaxAddCount, IStreamQueueCheckpointer<string> checkpointer, ICacheDataAdapter<EventData, TCachedMessage> cacheDataAdapter, ICacheDataComparer<TCachedMessage> comparer)
         {
-            cacheDataAdapter.PurgeAction = Purge;
+            this.defaultMaxAddCount = defaultMaxAddCount;
             Checkpointer = checkpointer;
-            OnPurged = CheckpointOnPurged;
+
+            cache = new PooledQueueCache<EventData, TCachedMessage>(cacheDataAdapter, comparer);
+            cacheDataAdapter.PurgeAction = cache.Purge;
+            cache.OnPurged = CheckpointOnPurged;
+
+            cachePressureMonitor = new CachePressureMonitor();
         }
 
+        protected abstract string GetOffset(TCachedMessage lastItemPurged);
+
+        // cachePressureContribution should be a double between 0-1, indicating how much danger the item is of being removed from the cache.  0 indicating  no danger, 1 indecating removal is imminent.
+        protected abstract bool TryCalculateCachePressureContribution(StreamSequenceToken token, out double cachePressureContribution);
+
         public void Dispose()
         {
-            OnPurged = null;
+            cache.OnPurged = null;
         }
 
-        protected abstract string GetOffset(TCachedMessage lastItemPurged);
-
         private void CheckpointOnPurged(TCachedMessage lastItemPurged)
         {
             Checkpointer.Update(GetOffset(lastItemPurged), DateTime.UtcNow);
         }
+
+        public int MaxAddCount { get { return cachePressureMonitor.IsUnderPressure() ? 0 : defaultMaxAddCount; } }
+
+        public void Add(EventData message)
+        {
+            cache.Add(message);
+        }
+
+        public object GetCursor(IStreamIdentity streamIdentity, StreamSequenceToken sequenceToken)
+        {
+            return cache.GetCursor(streamIdentity, sequenceToken);
+        }
+
+        public bool TryGetNextMessage(object cursorObj, out IBatchContainer message)
+        {
+            if (!cache.TryGetNextMessage(cursorObj, out message))
+                return false;
+            double cachePressureContribution;
+            if (TryCalculateCachePressureContribution(message.SequenceToken, out cachePressureContribution))
+            {
+                cachePressureMonitor.RecordCachePressureContribution(cachePressureContribution);
```

Comment:
if I understand correctly, you'll be recording values every time you read, and only once in a while averaging and resetting the recorded values. That could mean that a fast reader will record very often, and slow readers might not even record once before the values reset. Is that correct? Then the math is not exactly as explained, where each reader's weigth is the same as others. That is not necessarily wrong, as it might just mean that we need to lower/tune the threshold to account for it, but I'm pointing it out in case there is something a little bit better we could do, such as recording the value associated to each reader once (always overwriting the previous if there's any).